stdlib/mlang/symbolize.mc

```diff
@@ -0,0 +1,787 @@
+
+        let decl = DeclConDef {t with ident = ident,
+                                      tyIdent = symbolizeType env t.tyIdent} in 
+        let env = updateConEnv env conEnv in 
+        (env, decl)
+    | DeclUtest t -> 
+        -- This can be rewritten to use a shallow map on declarations. E.g.
+        -- smap (symbolizeExpr env) (DeclUtest t) 
+        let decl = DeclUtest {t with test = symbolizeExpr env t.test,
+                                     expected = symbolizeExpr env t.expected,
+                                     tusing = optionMap (symbolizeExpr env) t.tusing} in 
+        (env, decl)
+    | DeclExt t -> 
+        -- When ignoreExternals is set, the DeclExt should have been filtered
+        -- out of the program before attempting to symbolize the declarations.
+        if env.ignoreExternals then
+            error "DeclExt should have been removed when 'ignoreExternals' is true!"
+        else 
+            match setSymbol env.currentEnv.varEnv t.ident with (varEnv, ident) in
+            let decl = DeclExt {t with ident = ident,
+                                       tyIdent = symbolizeType env t.tyIdent} in 
+            let env = updateVarEnv env varEnv in 
+            (env, decl)
+    | DeclLang t -> 
+        -- Symbolize the name of the language
+        match setSymbol env.namespaceEnv t.ident with (namespaceEnv, ident) in 
+        let env = {env with namespaceEnv = namespaceEnv} in 
+
+        -- Symbolize included languages
+        let includes = map (getSymbol {kind = "language", info = [t.info], allowFree = false} env.namespaceEnv) t.includes in 
+
+        -- Create new langEnv and include the names defined in the included languages
+        let includedLangEnvs = map
+            (lam incl. match mapLookup (nameGetStr incl) env.langEnv 
+                       with Some langEnv then langEnv
+                       else errorMulti 
+                        [(t.info, nameGetStr incl)]
+                        "The included language can not be found!")
+            t.includes in 
+
+        let langEnv : NameEnv = foldl mergeNameEnv _nameEnvEmpty includedLangEnvs in 
+
+        let isSynDecl = lam d. match d with DeclSyn _ then true else false in 
+        let synDecls = filter isSynDecl t.decls in 
+
+        let isSemDecl = lam d. match d with DeclSem _ then true else false in 
+        let semDecls = filter isSemDecl t.decls in 
+
+        let isTypeDecl = lam d. match d with DeclType _ then true else false in 
+        let typeDecls = filter isTypeDecl t.decls in 
+
+        -- 1. Symbolize ident and params of SynDecls in this langauge
+        let symbSynStep1 = lam langEnv : NameEnv. lam synDecl.
+            match synDecl with DeclSyn s in
+            let env = updateEnv env langEnv in 
+
+            let ident = nameSym (nameGetStr s.ident) in 
+            match mapAccumL setSymbol env.currentEnv.tyVarEnv s.params with (_, params) in
+
+            let synn = DeclSyn {s with params = params,
```

Comment:
I feel like the first pass should only assign symbols to `ident`, because those will be in scope outside the `syn` itself, and then everything internal is symbolized after in a phase 2, much like how it's done for rec lets.

Reply:
Move symbolilzation of params to step 3.